### PR TITLE
prov/verbs: add FI_AV_TABLE support

### PR DESCRIPTION
**Pull request by evgeny-leksikov**
This series includes:
- add FI_AV_TABLE support
- refactoring of address vector and connection management. CM structure was moved to domain level to preserve not bound AV working

@a-ilango , please review. domain level is affected a bit

### CODE REVIEW

**Comment by sayantansur** [comment]
@a-ilango @evgeny-leksikov why isn't this done at a higher layer in the utility stuff? that way another provider that doesn't support FI_AV_TABLE natively can use this.

**Comment by a-ilango** [comment]
@sayantansur We have util_cmap interface in util_av to support FI_AV_TABLE. It's a more general purpose interface that allows storing connection management info and support FI_SOURCE capability. It supports FI_AV_TABLE implicitly.

The way it works is the user passes a handle to be stored in a cmap table during av_insert. The handle would be part of a structure that stores connection management info or FI_AV_MAP info relevant to the provider. During av_remove, util_cmap_free_handle, would remove the handle from the cmap table and also call any handler function (to close the connection for example) that was assigned during cmap initialization.

The cmap table itself is stored as part of the endpoint structure. @evgeny-leksikov If I understand correctly, the AV needs to be bound to an EP to be useful. So the connection management info has to stay with EP, right?

**Comment by sayantansur** [comment]
@evgeny-leksikov, any reason why the code that @a-ilango is pointing to cannot be used?

Less code is better!!

**Comment by evgeny-leksikov** [comment]
@a-ilango As I understand any av-operations (insert/remove) should work even without bound EP (av fabtest does it). But, you are right - it must be bound before real usage. On other hand (it looks like implementation specific), CM requires internal storage and progress thread. The idea of moving CM info into domain is also do not generate the overhead in case of several EPs per domain. It looks promising for easy implementation of scalable EP in the future as well.

@sayantansur Initial verbs/RDM implementation has many wrong things - for example, some fundamental things which should be encapsulated into special structures (AV, EP, domain, etc) are global. I'm working on refactoring in background to make it close to correct implementation. As soon as something can be easy merged with utils code - I'll do it.

**Comment by sayantansur** [comment]
@evgeny-leksikov sure, please proceed with the refactoring. It might be useful to structure the activity such that you and @a-ilango are in sync with the requirements.

**Comment by a-ilango** [comment]
@evgeny-leksikov even if there are several endpoints per domain, wouldn't they all have different transport addresses? A connection would be needed per transport address. That means the connection info has to stay with the endpoint. The CM progress thread and event handling can be at the domain level or even at the fabric level.

**Comment by evgeny-leksikov** [comment]
@a-ilango This is exactly that I noted in the code:

```
 +	/*
 +	 * TODO: Currently, only 1 rdm EP can be created per rdm domain!
 +	 *	 CM logic should be separated from EP,
 +	 *	 excluding naming/addressing
 +	 */

```
Connection info should stay with EP but connection management should be moved to upper level
I'll return address to EP and move the rest of CM logic to domain in next steps. I thought about fabric level too. If it will have any advantages - that shouldn't be a problem to move solid structure from domain to fabric but I need to implement through addressing of several endpoints before.

File: prov/verbs/src/fi_verbs.c [hunk 1]
```diff
@@ -71,6 +71,10 @@ static int fi_ibv_rdm_cm_init(struct fi_ibv_rdm_cm* cm,
 			      const struct rdma_addrinfo* rai)
 {
 	struct sockaddr_in* src_addr = (struct sockaddr_in*)rai->ai_src_addr;
+	if (cm->ec) {
+		/* cm is already initialized (by other ep) */
+		return FI_SUCCESS;
+	}
```
Comment (a-ilango) [inline prov/verbs/src/fi_verbs.c]: Accessing cm->ec may need to be serialized.

**Comment by a-ilango** [comment]
Sorry, I didn't notice the TODO. The PR looks good except for the one comment on accessing cm->ec. :+1: 
Let me know when this is ready for merge.

**Comment by evgeny-leksikov** [comment]
@a-ilango I've moved addr info back to EP, Moving of CM thread and event handling to domain requires significant changes. So, I'll create a new PR later. 
If you are OK with changes, it's ready for merge.

**Pull request merged**